### PR TITLE
Initial NodeValueReader Implementation

### PR DESCRIPTION
**Pull request by simpsont-oci**
Problem: Currently, the node-opendds plugin uses a mish-mash of serialization & deserialization styles. Conversion from C++ IDL objects to JSON uses the NodeValueWriter, but conversion from JSON objects to C++ IDL uses the v8_generator generated type support.

Solution: Add NodeValueReader and move the JSON-to-C++ conversion to use this interface. This should also allow the future removal of the generated v8 typesupport and the (klunky) v8stubs library currently required for testing.

### CODE REVIEW

File: src/NodeValueReader.h [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+private:
+  NodeValueReader();
+
+  std::string indent(int mod);
+
+  bool begin_nested();
+  bool end_nested();
+
+  template <typename V, typename T>
+  bool primitive_helper(T& value, bool (v8::Value::*checker)() const)
```
Comment (mitza-oci) [inline src/NodeValueReader.h:87]: Using typedefs to give these "pointer-to-member function" and function pointer types their own names may improve readability (here and below).

File: src/node-opendds.cpp [hunk 2]
```diff
@@ -432,15 +422,41 @@ namespace {
...
       return;
     }
+
     void* const dw_i = Nan::GetInternalFieldPointer(fci.This(), 0);
-    const void* const tc_i = Nan::GetInternalFieldPointer(fci.This(), 1);
     DDS::DataWriter* const dw = static_cast<DDS::DataWriter*>(dw_i);
-    const OpenDDS::DCPS::V8TypeConverter* const tc = reinterpret_cast<const OpenDDS::DCPS::V8TypeConverter*>(tc_i);
+
+    const void* const ts_i = Nan::GetInternalFieldPointer(fci.This(), 1);
+    const OpenDDS::DCPS::TypeSupport* const ts = reinterpret_cast<const OpenDDS::DCPS::TypeSupport*>(ts_i);
```
Comment (mitza-oci) [inline src/node-opendds.cpp:430]: Does this need to be a reinterpret?  A typed pointer that goes to `void*` and then comes back can use `static_cast`.

File: src/NodeValueReader.cpp [hunk 3]
```diff
@@ -0,0 +1,487 @@
...
+  }
+  return result;
+}
+
+} // namespace
+
+NodeValueReader::NodeValueReader(v8::Local<v8::Object> obj)
+{
+  indent_ = 0;
+  //std::cout << indent(0) << "NodeValueReader::NodeValueReader() - " << json_stringify(obj) << std::endl;
```
Comment (mitza-oci) [inline src/NodeValueReader.cpp:29]: remove?

File: src/NodeValueReader.cpp [hunk 4]
```diff
@@ -0,0 +1,487 @@
...
+    //    value = static_cast<ACE_CDR::LongLong>(tov.ToLocalChecked()->Int64Value());
+    //    return true;
+    //  }
+    //}
+    {
+      Nan::MaybeLocal<v8::String> tov = Nan::To<v8::String>(mlvai.ToLocalChecked());
+      if (!tov.IsEmpty()) {
+        std::string temp;
+        read_string(temp);
+#ifndef ACE_LINUX
```
Comment (mitza-oci) [inline src/NodeValueReader.cpp:277]: Is there a more specific macro we could use?  This doesn't seem to be related to the OS (for example, different compilers on the same OS may do this differently).

File: src/NodeValueReader.cpp [hunk 5]
```diff
@@ -0,0 +1,487 @@
...
+  }
+  return false;
+}
+
+bool NodeValueReader::read_long_enum(ACE_CDR::Long& value, const OpenDDS::DCPS::EnumHelper& helper)
+{
+  //std::cout << indent(0) << "NodeValueReader::read_long_enum()" << std::endl;
+  Nan::MaybeLocal<v8::Value> mlvai = use_name_ ? Nan::Get(current_object_, current_property_name_) : Nan::Get(current_object_, current_index_);
+  if (!mlvai.IsEmpty()) {
+    {
```
Comment (jrw972) [inline src/NodeValueReader.cpp:417]: Eliminate extra scope.

**Review by jrw972 — CHANGES_REQUESTED** [review]
Clean up scaffolding.

**Review by mitza-oci — COMMENTED** [review]
There's a bunch of warnings like this
`unused variable 'DDSSEC_PROP_IDENTITY_CA' [-Wunused-const-variable]`

File: src/NodeValueReader.h (hunk 1 above)
Comment (mitza-oci) [reply src/NodeValueReader.h]: It looks like there's more of these that could change.  Doesn't need to happen right away this PR.

Comment (simpsont-oci) [reply src/NodeValueReader.h]: I'm not sure the str_conv ones can get a simple typedef because their return types depend on a template parameter.

**Review by mitza-oci — APPROVED** [review]
